src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,166 @@
+
+    internal static bool CanRewrite(CSharpCompilation compilation, MethodSymbol topLevelMethod, bool inExpressionLambda, BoundConversion boundConversion, MethodSymbol targetMethod)
+        => targetMethod.IsStatic
+        && !boundConversion.IsExtensionMethod
+        && !inExpressionLambda // The tree structure / meaning for expression trees should remain untouched.
+        && topLevelMethod.MethodKind != MethodKind.StaticConstructor // Avoid caching twice if people do it manually.
+        && compilation.IsStaticMethodGroupDelegateCacheEnabled
+        ;
+
+    internal BoundExpression Rewrite(int localFunctionOrdinal, SyntaxNode syntax, BoundExpression receiver, MethodSymbol targetMethod, TypeSymbol delegateType)
+    {
+        Debug.Assert(delegateType.IsDelegateType());
+
+        var oldSyntax = _factory.Syntax;
+        _factory.Syntax = syntax;
+
+        var cacheContainer = GetOrAddCacheContainer(localFunctionOrdinal, delegateType, targetMethod);
+        var cacheField = cacheContainer.GetOrAddCacheField(_factory, delegateType, targetMethod);
+
+        var boundCacheField = _factory.Field(null, cacheField);
+        var boundDelegateCreation = new BoundDelegateCreationExpression(syntax, receiver, targetMethod, isExtensionMethod: false, type: delegateType);
+
+        var rewrittenNode = _factory.Coalesce(boundCacheField, _factory.AssignmentExpression(boundCacheField, boundDelegateCreation));
+
+        _factory.Syntax = oldSyntax;
+
+        return rewrittenNode;
+    }
+
+    private DelegateCacheContainer GetOrAddCacheContainer(int localFunctionOrdinal, TypeSymbol delegateType, MethodSymbol targetMethod)
+    {
+        Debug.Assert(_factory.ModuleBuilderOpt is { });
+        Debug.Assert(_factory.CurrentFunction is { });
+
+        var typeCompilationState = _factory.CompilationState;
+        var generation = _factory.ModuleBuilderOpt.CurrentGenerationOrdinal;
+
+        DelegateCacheContainer? container;
+
+        if (CanUseConcreteCacheContainer(delegateType, targetMethod))
+        {
+            container = typeCompilationState.ConcreteDelegateCacheContainer;
+
+            if (container is { })
+            {
+                return container;
+            }
+
+            container = new(typeCompilationState.Type, generation);
+            typeCompilationState.ConcreteDelegateCacheContainer = container;
+        }
+        else
+        {
+            var containersStack = _genericCacheContainers ??= new();
+
+            while (containersStack.Count > 0 && containersStack.Peek().LocalFunctionOrdinal > localFunctionOrdinal)
+            {
+                containersStack.Pop();
+            }
+
+            if (containersStack.Count > 0 && containersStack.Peek().LocalFunctionOrdinal == localFunctionOrdinal)
+            {
+                return containersStack.Peek().CacheContainer;
+            }
+
+            container = new(_factory.CurrentFunction, _topLevelMethodOrdinal, localFunctionOrdinal, generation);
```

Comment:
>new(_factory.CurrentFunction, _topLevelMethodOrdinal, localFunctionOrdinal, generation); [](http://example.com/codeflow?start=24&length=88)

It feels unnecessary to allocate a container for a local function that itself doesn't have type parameters. I think we only need containers for functions that have type parameters. Optionally, we might even decide to skip some containers because they are associated with type parameters that are not significant for the specific delegate creation.

Reply:
I can't quite get you, could you please clarify? If you're talking about targeting local functions that does not declare type parameters in its own declarations, please check https://github.com/dotnet/roslyn/pull/58288/commits/1daafd8da97ef33dc87af7cd8fb16ba8d952abd0

Reply:
> I can't quite get you, could you please clarify?

Consider the following scenario:
```
void M()
{
    void LF1<T>()
    {
         void LF2()
         {
               Func<T> d = SomeMethod<T>;
         }

         void LF3()
         {
               Func<T> d = SomeMethod<T>;
         }
    }
}
```

There is no need to create container for LF2 and LF3. Both delegate creations can be served with container for LF1. In fact the cache can be shared across both delegate creations, which is likely to happen automatically if we share the container.

Reply:
Brilliant.

Reply:
Please check 9ec9776